### PR TITLE
[WIP] Added async framework.

### PR DESCRIPTION
**Pull request by Thom1729**
This is a first pass at implementing the ideas outlined in #90.

The core function is `run(function, scheduler)`. `scheduler` may be `set_timeout`, `set_timeout_async`, or any function with the same signature. Two new such schedulers are provided. `thread_scheduler()` starts a new thread, runs the task, and ends the thread. `SharedThreadRunner` maintains a shared thread in which parts of tasks are enqueued.

I think that the core should be fairly solid. The API is a work in progress.

Example tasks:

```python
def delay_task():
    for i in range(5):
        yield 500 # Wait 500 ms
        print(i)

from sublime_lib import show_selection_panel
def selection_task():
    try:
        result = yield promise(lambda future:
            show_selection_panel(
                sublime.active_window(),
                ['Alice', 'Bob', 'Carol'],
                on_select=future.set_result,
                on_cancel=lambda: future.set_exception(ValueError("Cancelled"))
            )
        )
    except ValueError as ex:
        print(ex)
    else:
        print(result)
```

Invocations:

```python
run(delay_task, sublime.set_timeout)
run(delay_task, thread_scheduler())

shared = SharedThreadRunner()
run(delay_task, shared.schedule)
```

### CODE REVIEW

File: st3/sublime_lib/async.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    def decorator(function):
+        return lambda *args, **kwargs: run(partial(function, *args, **kwargs))
+
+    return decorator
+
+
+    from functools import partial
+    from threading import Thread, Event
+    from time import sleep, monotonic
+    from queue import Queue, PriorityQueue, Empty
```
Comment (FichteFoll) [inline st3/sublime_lib/async.py]: What's this block?

File: st3/sublime_lib/async.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                action.add_done_callback(
+                    lambda f: self._schedule(partial(self._next, f), 0)
+                )
+            else:
+                raise TypeError("Task returned invalid action.")
+
+
+def async_decorator(run):
+    def decorator(function):
+        return lambda *args, **kwargs: run(partial(function, *args, **kwargs))
```
Comment (FichteFoll) [inline st3/sublime_lib/async.py]: I guess this should be a proper function wrapped with `functools.wraps`.

File: st3/sublime_lib/async.py (hunk 1 above)
Comment (Thom1729) [reply st3/sublime_lib/async.py]: In the wrong place. Fixed.

File: st3/sublime_lib/async.py (hunk 2 above)
Comment (Thom1729) [reply st3/sublime_lib/async.py]: Fixed.

**Comment by Thom1729** [comment]
Added [WIP] label to clarify that the API is not great and I don't think it's quite ready for release.

**Comment by Thom1729** [comment]
Since Python 3.8 provides `async`, this probably isn't necessary.

**Pull request closed without merging**